File: src/java/ironfist/Creature.java
package ironfist;

import ironfist.math.Vector2D;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * DOCUMENT ME!
 * 
 * @author pmgremo
 */
public class Creature implements Serializable {

  private static final long serialVersionUID = 3257003250497630515L;

  private static final RecursiveShadowCastingVision fov = new RecursiveShadowCastingVision();

  private Map<Object, Object> properties;
  private String name;
  private Vector2D coordinate;
  private Client client;
  private boolean levelChanged;
  private Level level;
  private List<Thing> things;
  private int power;

  {
    properties = new HashMap<Object, Object>();
    things = new LinkedList<Thing>();
  }

  /**
   * Creates a new Creature object.
   * 
   * @param client
   *          DOCUMENT ME!
   */
  public void setClient(Client client) {
    this.client = client;
  }

  /**
   * Returns the coordinate.
   * 
   * @return Coordinate
   */
  public Vector2D getCoordinate() {
    return coordinate;
  }

  /**
   * Sets the coordinate.
   * 
   * @param coordinate
   *          The coordinate to set
   */
  public void setCoordinate(Vector2D coordinate) {
    this.coordinate = coordinate;
  }

  /**
   * DOCUMENT ME!
   */
  public synchronized void activate() {
    if (levelChanged) {
      client.onLevelChange(level);
      levelChanged = false;
    }

    client.onVisionChange(getVision());

    power = 1;

    while (power > 0) {
      try {
        wait();
      } catch (InterruptedException e) {
      }
    }
  }

  /**
   * DOCUMENT ME!
   * 
   * @param command
   *          DOCUMENT ME!
   */
  public synchronized void executeCommand(Command command) {
    CommandType type = command.getType();

    if (CommandType.MOVE.equals(type)) {
      Referee.move(this, (Vector2D) command.getParameter());
    } else if (CommandType.OPEN.equals(type)) {
      Referee.open(this, (Door) command.getParameter());
    } else if (CommandType.CLOSE.equals(type)) {
      Referee.close(this, (Door) command.getParameter());
    } else if (CommandType.UP.equals(type)) {
      Referee.up(this);
    } else if (CommandType.DOWN.equals(type)) {
      Referee.down(this);
    } else if (CommandType.QUIT.equals(type)) {
      Referee.quit(this);
    } else if (CommandType.SAVE.equals(type)) {
      Referee.save(this);
    } else if (CommandType.PICKUP.equals(type)) {
      Referee.pickup(this, (Thing) command.getParameter());
    } else if (CommandType.DROP.equals(type)) {
      Referee.drop(this, (Thing) command.getParameter());
    }

    power--;
    notify();
  }

  /**
   * Returns the level.
   * 
   * @return Level
   */
  public Level getLevel() {
    return level;
  }

  /**
   * Sets the level.
   * 
   * @param level
   *          The level to set
   */
  public void setLevel(Level level) {
    this.level = level;
    levelChanged = true;
  }

  /**
   * Returns the name.
   * 
   * @return String
   */
  public String getName() {
    return name;
  }

  /**
   * Sets the name.
   * 
   * @param value
   *          The name to set
   */
  public void setName(String value) {
    this.name = value;
  }

  /**
   * DOCUMENT ME!
   * 
   * @param key
   *          DOCUMENT ME!
   * 
   * @return DOCUMENT ME!
   */
  public Object getProperty(Object key) {
    return properties.get(key);
  }

  /**
   * DOCUMENT ME!
   * 
   * @param key
   *          DOCUMENT ME!
   * @param value
   *          DOCUMENT ME!
   */
  public void setProperty(Object key, Object value) {
    properties.put(key, value);
  }

  /**
   * DOCUMENT ME!
   * 
   * @param key
   *          DOCUMENT ME!
   */
  public void removeProperty(Object key) {
    properties.remove(key);
  }

  /**
   * DOCUMENT ME!
   * 
   * @param thing
   *          DOCUMENT ME!
   */
  public void addThing(Thing thing) {
    things.add(thing);
  }

  /**
   * DOCUMENT ME!
   * 
   * @param thing
   *          DOCUMENT ME!
   */
  public void removeThing(Thing thing) {
    things.remove(thing);
  }

  /**
   * DOCUMENT ME!
   * 
   * @return DOCUMENT ME!
   */
  public Iterator<Thing> getThings() {
    return things.iterator();
  }

  /**
   * DOCUMENT ME!
   * 
   * @return DOCUMENT ME!
   */
  public List<Tile> getVision() {
    Vector2D location = getCoordinate();
    return fov.getSeen(getLevel(), (int) location.getX(),
      (int) location.getY(), 3);
  }
}

File: src/main/java/ukalus/Creature.kt
package ukalus

import ukalus.math.Vector2D

import java.io.Serializable
import java.util.HashMap
import java.util.LinkedList

/**
 * DOCUMENT ME!

 * @author pmgremo
 */
class Creature : Serializable {
    private val lock = java.lang.Object()

    private var properties: MutableMap<Any, Any>? = null
    /**
     * Returns the name.

     * @return String
     */
    /**
     * Sets the name.

     * @param value
     * *          The name to set
     */
    var name: String? = null
    /**
     * Returns the coordinate.

     * @return Coordinate
     */
    /**
     * Sets the coordinate.

     * @param coordinate
     * *          The coordinate to set
     */
    var coordinate: Vector2D? = null
    private var client: Client? = null
    private var levelChanged: Boolean = false
    /**
     * Returns the level.

     * @return Level
     */
    /**
     * Sets the level.

     * @param level
     * *          The level to set
     */
    var level: Level? = null
        set(level) {
            field = level
            levelChanged = true
        }
    private var things: MutableList<Thing>? = null
    private var power: Int = 0

    init {
        properties = HashMap<Any, Any>()
        things = LinkedList<Thing>()
    }

    /**
     * Creates a new Creature object.

     * @param client
     * *          DOCUMENT ME!
     */
    fun setClient(client: Client) {
        this.client = client
    }

    /**
     * DOCUMENT ME!
     */
    @Synchronized fun activate() {
        if (levelChanged) {
            client!!.onLevelChange(this.level!!)
            levelChanged = false
        }

        client!!.onVisionChange(vision)

        power = 1

        while (power > 0) {
            try {
                lock.wait()
            } catch (e: InterruptedException) {
            }

        }
    }

    /**
     * DOCUMENT ME!

     * @param command
     * *          DOCUMENT ME!
     */
    @Synchronized fun executeCommand(command: Command) {
        val type = command.type

        when (type) {
            CommandType.MOVE -> Referee.move(this, command.parameter as Vector2D)
            CommandType.OPEN -> Referee.open(this, command.parameter as Door?)
            CommandType.CLOSE -> Referee.close(this, command.parameter as Door?)
            CommandType.UP -> Referee.up(this)
            CommandType.DOWN -> Referee.down(this)
            CommandType.QUIT -> Referee.quit(this)
            CommandType.SAVE -> Referee.save(this)
            CommandType.PICKUP -> Referee.pickup(this, command.parameter as Thing)
            CommandType.DROP -> Referee.drop(this, command.parameter as Thing)
        }

        power--
        lock.notify()
    }

    /**
     * DOCUMENT ME!

     * @param key
     * *          DOCUMENT ME!
     * *
     * *
     * @return DOCUMENT ME!
     */
    fun getProperty(key: Any): Any? {
        return properties!![key]
    }

    /**
     * DOCUMENT ME!

     * @param key
     * *          DOCUMENT ME!
     * *
     * @param value
     * *          DOCUMENT ME!
     */
    fun setProperty(key: Any, value: Any) {
        properties!!.put(key, value)
    }

    /**
     * DOCUMENT ME!

     * @param key
     * *          DOCUMENT ME!
     */
    fun removeProperty(key: Any) {
        properties!!.remove(key)
    }

    /**
     * DOCUMENT ME!

     * @param thing
     * *          DOCUMENT ME!
     */
    fun addThing(thing: Thing) {
        things!!.add(thing)
    }

    /**
     * DOCUMENT ME!

     * @param thing
     * *          DOCUMENT ME!
     */
    fun removeThing(thing: Thing) {
        things!!.remove(thing)
    }

    /**
     * DOCUMENT ME!

     * @return DOCUMENT ME!
     */
    fun getThings(): Iterator<Thing> {
        return things!!.iterator()
    }

    /**
     * DOCUMENT ME!

     * @return DOCUMENT ME!
     */
    val vision: List<Tile>
        get() = if (coordinate == null) emptyList() else fov.getSeen(level, coordinate!!.x, coordinate!!.y, 3)

    companion object {

        private const val serialVersionUID = 3257003250497630515L

        private val fov = RecursiveShadowCastingVision()
    }
}